scratch/benchmarks/benchmark_fgpu.py

```diff
@@ -322,14 +344,23 @@ async def compare(adc_sample_rate: float) -> bool:
     if high_result.good():
         raise RuntimeError("succeeded on high")
 
-    # Compute error estimate. This is a heuristic based on gut feel rather
-    # than measurement. Assume that results are 1-NOISE reliable when at least
-    # FUZZ away from the critical value, and vary logarithmically in between.
+    # Compute error estimate. The model is a logistic regression on the
+    # log of the sample rate (log is used mainly to make the slope invariant
+    # to the scale of the rates, rather than for the shape).
+    # The magic numbers are determined from fit.py.
+    slope = {
+        1: -342.212919,
+        2: -173.264274,
+        4: -582.668296,
+    }[args.n]
     mid_rates = 0.5 * (rates[:-1] + rates[1:])  # Rates in the middle of the intervals
     mid_rates = np.r_[args.low, mid_rates, args.high]
     l_rates = np.log(rates)[:, np.newaxis]
     l_mid_rates = np.log(mid_rates)[np.newaxis, :]
-    noise = np.clip(0.5 + (l_rates - l_mid_rates) / np.log1p(FUZZ) * (0.5 - NOISE), NOISE, 1 - NOISE)
+    noise = expit((l_mid_rates - l_rates) * slope)
```

Comment:
I'm not very clear on this noise term... I thought the expit would represent the probability of success at various (log) data rates, so that for a certain rate r0 we have P = 0.5 (the transition rate) and for r << r0 we have near certain success and r >> r0 we have near certain failure.

So why are you interested in the ratio of the "mid" rate and the rate below (?) it and passing that to expit instead?

Reply:
l_rates represents the rates at which we might choose to run trials. But the probability distribution is discretised as the probabilities that the transition rate is in each interval between two of those trial rates. Rather than mess around with integration, I'm just using the mid-point of each interval as a proxy for the whole interval.

Note that l_rates is a column vector and l_mid_rates is a row vector, so you're getting broadcasting here. In your notation, noise[i, j] is the probability of success (or failure, I forget which offhand) if log(r0) is l_mid_rates[j] and log(r) is l_rates[i].

I put everything into log-space just to make it scale-invariant e.g. the `slope` term shouldn't need to change if you run this on a system that's twice as fast or if you change the units of rate from Hz to MHz.

Reply:
Thanks, I missed the broadcast!